### PR TITLE
drivers: rcar_mmc: add request retry mechanism

### PR DESCRIPTION
**Pull request by xakep-amatop**
Add request retry mechanism to Renesas SD/MMC driver.

### CODE REVIEW

File: drivers/sdhc/rcar_mmc.c [hunk 1]
```diff
@@ -744,72 +744,81 @@ static int rcar_mmc_request(const struct device *dev,
...
 	bool is_read = true;
+	int retries;
 
 	if (!dev || !cmd) {
 		return -EINVAL;
 	}
 
-	/* TODO: add retries mechanism, use retries counter from cmd for that */
 	response_type = cmd->response_type & SDHC_NATIVE_RESPONSE_MASK;
+	retries = cmd->retries ? : 1;
```
Comment (lorc) [inline drivers/sdhc/rcar_mmc.c]: This does not look right. In my understanding, `retries` tells how many times you should retry operation if it fails. So, for example, `cmd->retries == 0` tells that you should not retry operation at all and `cmd->retries == 1` tells you that you should try one more time. At least, this is how all other SDHC drivers behave. 

But your code makes this two cases equal.

File: drivers/sdhc/rcar_mmc.c [hunk 2]
```diff
@@ -744,72 +744,81 @@ static int rcar_mmc_request(const struct device *dev,
...
 
-	if (rcar_mmc_card_busy(dev)) {
-		LOG_ERR("%s: device busy", dev->name);
-		return -EBUSY;
-	}
+	while (ret && retries-- > 0) {
+		ret = rcar_mmc_poll_reg_flags_check_err(dev, RCAR_MMC_INFO2,
+							RCAR_MMC_INFO2_CBSY, 0, false, false);
+		if (ret) {
+			LOG_ERR("%s: device busy", dev->name);
```
Comment (lorc) [inline drivers/sdhc/rcar_mmc.c]: If upper layer configures number of retries, it expects that there may be an recoverable error. So I believe, you don't need to report each error with the `ERR` print. Other drivers use `LOG_DBG` in a such case.

**Review by lorc — COMMENTED** [review]
Please provide a comprehensive commit description.

File: drivers/sdhc/rcar_mmc.c (hunk 1 above)
Comment (xakep-amatop) [reply drivers/sdhc/rcar_mmc.c]: done

File: drivers/sdhc/rcar_mmc.c (hunk 2 above)
Comment (xakep-amatop) [reply drivers/sdhc/rcar_mmc.c]: done

> Please provide a comprehensive commit description.

done

**Comment by lorc** [comment]
Thanks!

>  However, can be some types of errors after which it isn't necessary to retry

Word missing? Did you wanted to say "However, there can be some types..."?

With that fixed:

`Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`

**Review by firscity — APPROVED** [review]
`Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>`

**Pull request merged**